pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,334 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package ebpf
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+)
+
+var (
+	batchAPISupportedOnce sync.Once
+	batchAPISupported     bool
+)
+
+// BatchAPISupported returns true if the kernel supports the batch API for maps
+func BatchAPISupported() bool {
+	batchAPISupportedOnce.Do(func() {
+		// Do feature detection directly instead of based on kernel versions for more accuracy
+		m, err := ebpf.NewMap(&ebpf.MapSpec{
+			Type:       ebpf.Hash,
+			KeySize:    4,
+			ValueSize:  4,
+			MaxEntries: 10,
+		})
+		if err != nil {
+			log.Warnf("Failed to create map for batch API test: %v, will mark batch API as unsupported", err)
+			batchAPISupported = false
+			return
+		}
+
+		keys := make([]uint32, 1)
+		values := make([]uint32, 1)
+
+		// Do a batch update, check the result.
+		// We do an update instead of a lookup because it's more reliable for detection
```

Comment:
The initial version of this check had some calls to `Put()` and then a `BatchLookup` call. Sometimes it returned an error or returned no elements, I couldn't reproduce it well (and it was probably due to a mistake on my part rather than in the ebpf library) so I decided to use the `BatchUpdate` call, it was introduced in the same changeset and I think in any case it's better because it's just one operation that has less reasons to fail than the lookup.